[IR] llvm.reloc.none intrinsic for no-op symbol references

mysterymath opened the pull request:
This intrinsic emits a BFD_RELOC_NONE relocation at the point of call, which allows optimizations and languages to explicitly pull in symbols from static libraries without there being any code or data that has an effectual relocation against such a symbol.

See issue #146159 for context.

mysterymath committed:
[IR] llvm.reloc.none intrinsic for no-op symbol references

This intrinsic emits a BFD_RELOC_NONE relocation at the point of call,
which allows optimizations and languages to explicitly pull in symbols
from static libraries without there being any code or data that has an
effectual relocation against such a symbol.

See issue #146159 for context.

mysterymath commented:
Sending this out as a draft to obtain some early feedback about the direction of the implemenation of the Modular Printf RFC.

petrhosek commented on llvm/docs/LangRef.rst:
```diff
@@ -30439,6 +30439,38 @@ This intrinsic does nothing, but optimizers must consider it a use of its single
+Arguments:
+""""""""""
+
+The ``llvm.fake.use`` intrinsic takes one argument, which may be any global
```
```suggestion
The ``llvm.reloc.none`` intrinsic takes one argument, which may be any global
```

mysterymath replied:
Done.

mysterymath commented on llvm/docs/LangRef.rst:
```diff
@@ -30439,6 +30439,38 @@ This intrinsic does nothing, but optimizers must consider it a use of its single
+Overview:
+"""""""""
+
+The ``llvm.reloc.none`` intrinsic emits a no-op relocation against a given
+operand symbol. This can bring the symbol
+definition into the link without emitting any code or data to the binary for
+that purpose.
```
The reason being... I had no idea you could do that! This is essentially my first time monkeying in the IR; I've been a backend/linker jocky for the most part. But thanks, that does seem a lot semantically simpler. I'll make the change throughout the PR chain and report back here.

mysterymath committed:
fake.use -> reloc.none

mysterymath committed:
Take symbol name by metadata arg rather than ptr to GlobalValue

mysterymath committed:
Add a generic reloc_none test

mysterymath committed:
IR verifier check and test

mysterymath commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -7757,6 +7757,19 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
+    SDValue V = getValue(I.getArgOperand(0));
+    auto *GA = dyn_cast<GlobalAddressSDNode>(V);
+    if (!GA)
+      report_fatal_error("llvm.reloc.none operand must be a GlobalValue");
```
Thanks; done.

arsenm commented on llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp:
```diff
@@ -1930,6 +1930,20 @@ void AsmPrinter::emitFunctionBody() {
+        OutStreamer->emitLabel(Sym);
+        const MCExpr *Dot = MCSymbolRefExpr::create(Sym, OutContext);
+
+        assert(MI.getNumOperands() == 1 &&
+               "RELOC_NONE can only have one operand");
```
Belongs in MachineVerifier

arsenm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -7757,6 +7757,20 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
+  case Intrinsic::reloc_none: {
+    Metadata *MD = cast<MetadataAsValue>(I.getArgOperand(0))->getMetadata();
+    StringRef SymbolName = cast<MDString>(MD)->getString();
+    auto *M = const_cast<Module *>(I.getModule());
+    auto *RelocSymbol = cast<GlobalVariable>(
+        M->getOrInsertGlobal(SymbolName, StructType::create(M->getContext())));
```
You cannot modify the IR here

mysterymath replied:
This is because this occurs in a `MachineFunctionPass` right? In that case, I got bamboozled! I had unceremoniously copy-pasted this from the lowering of `Intrinsic::amdgcn_reloc_constant` in `SIISelLowering.cpp`, which also runs in a `MachineFunctionPass`. I'll admit that this completely slipped past me, although I was curious about the `const_cast`.

AFAIK, converting an arbitrary symbol name metadata string to a GlobalValue reference instrinsically mutates the module. So, if we wanted to fix this, we could either go back to having this be a symbol reference to a pre-existing global value, or we could  forward the metadata string to `RELOC_NONE` as a string (odd in the backend; unsure if this is even possible?) to be lowered to a symbol reference at assembly time. My preference would definitely be the former, since it has fewer unknowns.

Whichever we choose, also I'd think either way we should give `amdgcn_reloc_constant` the same treatment, unless it's not possible to do so for backwards compatibility reasons.

@arsenm Do you have preferences here?

arsenm replied:
You don't need a GlobalValue reference to emit the symbol name in the end. You just need to encode this as an ExternalSymbol operand instead. The comment that it's "for debug info" doesn't really matter

arsenm commented on llvm/test/Verifier/reloc_none.ll:
```diff
@@ -0,0 +1,13 @@
+; RUN: not opt -S -passes=verify 2>&1 < %s | FileCheck %s
```
```suggestion
; RUN: not llvm-as -disable-output 2>&1 %s | FileCheck %s
```

mysterymath committed:
Remove unneeded assertion from AsmPrinter

mysterymath committed:
Use llvm-as for test

mysterymath committed:
Rename reloc_none.ll to reloc-none.ll

mysterymath committed:
Lower reloc.none in Global ISel

mysterymath committed:
Remove arg from emitRelocDirective call

mysterymath committed:
Update tests

mysterymath commented:
Alright, I've filled out tests and added lowering to GlobalISel also. This should be formally ready for review (although much has already occurred).

mysterymath committed:
Take symbol name by GlobalValue again to avoid modifying Module

MaskRay commented on llvm/include/llvm/IR/Intrinsics.td:
```diff
@@ -1913,6 +1913,9 @@ def int_threadlocal_address : DefaultAttrsIntrinsic<[llvm_anyptr_ty], [LLVMMatch
                                             [], [IntrNoMem]>;
 
+def int_reloc_none : DefaultAttrsIntrinsic<[], [llvm_ptr_ty],
+  [IntrHasSideEffects, IntrInaccessibleMemOnly, IntrWillReturn]>;
```
Why define IntrHasSideEffects?

arsenm replied:
New uses of llvm_ptr_ty should not be introduced. Only llvm_anyptr_ty

mysterymath committed:
Take argument once again as a metadata string.

This reverts commit b01bf98300b537bb0e5ccea42d7735ffd0c02e24.

However, this time, the metadata string is passed as a string directly
to the backend. This prevents the affected MachineFunction passes from
incorrectly modifying the IR.

mysterymath committed:
Remove cargo cult attribute; test for no DCE at O2

mysterymath commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -7750,6 +7750,17 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
+    SDValue Ops[2];
+    Ops[0] = getRoot();
+    Ops[1] = DAG.getTargetGlobalAddress(GA->getGlobal(), sdl, V.getValueType(),
+                                        GA->getOffset());
```
Thanks; done.

mysterymath committed:
Arsenm style suggestion

mysterymath committed:
Fix langref formatting

mysterymath reviewed (commented):
Finally got back around to this one. I've modified the backend pseudo to use an `ExternalSymbol` instead and the IR intrinsic to take metadata once gain; this ended up much cleaner, as well as fixing the invariant violation. I had originally tried this, but I had thought from the interface that ExternalSymbol could only take a constant string or something interned and null terminated. Thankfully, metadata strings are interned and null terminated, although it was surprisingly difficult to figure that out.

That should address the outstanding comments; PTAL and let me know if there's anything  else.

mysterymath committed:
Update newly-broken tests

mysterymath committed:
Add side effect tag, since opt (not llc) will strip reloc_none

mysterymath commented:
Coming back from the US dev meeting, I wanted to ping this one again. Is there anything left for me to do before this is ready to land?

ilovepi reviewed (approved):
Basically LGTM. I see a comment mentioning testing for robustness against DCE ,but I don't see that in the remaining tests ... was this present in an earlier version of the patch?

please check w/ other reviewers before landing. In particular, I see @MaskRay had a question that seems worth answering/discussing.

arsenm commented on llvm/docs/LangRef.rst:
```diff
@@ -30947,6 +30947,37 @@ This intrinsic does nothing, but optimizers must consider it a use of its single
+
+::
+
+      declare void @llvm.reloc.none(ptrty %ptr)
```
Signature doesn't match the new metadata version

mysterymath replied:
Ah thanks; done.

mysterymath committed:
Correct argument in docs

mysterymath commented:
From @ilovepi:
> In particular, I see @MaskRay had a question that seems worth answering/discussing.

Ah, thanks.

From @MaskRay:
> Note: If a function A with llvm.reloc.none is inlined into other functions, the section referenced by llvm.reloc.none becomes harder for the linker to garbage collect.

Unless I'm misunderstanding, this seems desirable. Presumably code that bears this intrinsic intends to declare that the referenced symbol is needed for something; I would expect such a declaration to be reasonably sticky. For the modular printf itself, these symbols will probably be zero-length bearers of additional relocations to pull in the actual implementation functions. In such cases too, keeping the chain of symbols/sections around makes it more obvious why things are present, a la `--why-extract` and `--why-live`.

> As an intrinsic, this cannot be used with global objects. The !associated metadata can serve as a replacement for ELF in some cases.

Ack; I do like that there are options to generalize this beyond code locations if necessary, but it does seem sensible for there to be more than one spelling too, given the strong-ish split between code and data in LLVM IR. I suppose we could try to make this attribute-only in code too, but that seems... unusual.

mysterymath committed:
Update ir2vec goldens

mysterymath commented:
> The tests in this patch hang when running on Hexagon during VLIW scheduling. It looks like setting hasSideEffects = true in Target.td causes the Hexagon scheduler to enter an infinite loop. @mysterymath Could you please take a look?

Can you provide a bit more detail about how I can reproduce this?
